### PR TITLE
DOC: allow np as a module for documentation

### PR DESCRIPTION
**Pull request by mattip**
Add `np` to sys.modules so that documentation can use `` `np.dtype` `` instead of `` `numpy.dtype` `` or `` `np.dtype <numpy.dtype>`

### CODE REVIEW

**Comment by mattip** [comment]
hmm, segfault when building the documentation.

**Comment by eric-wieser** [comment]
Does the following work better:
```
aliases = {}
for name, module in sys.modules.items():
    parts = name.split('.', 1)
    if parts[0] == 'numpy':
        parts[0] = 'np'
        aliases[name] = module
sys.modules.update(aliases)
```

**Comment by mattip** [comment]
It still may fail if new numpy c-extension imports are needed past this point, which should not be the case in documentation.

File: doc/source/conf.py [hunk 1]
```diff
@@ -41,10 +41,21 @@
...
+import numpy
+
+# Allow the shortcut `np.dtype`. 
+# May fail if the shortcut needs to newly import.
+aliases = {}
+for name, module in sys.modules.items():
+    parts = name.split('.', 1)
+    if parts[0] == 'numpy':
+        parts[0] = 'np'
+        aliases[name] = module
```
Comment (eric-wieser) [inline doc/source/conf.py]: I missed a line, sorry: `name = '.'.join(parts)`

**Comment by mattip** [comment]
Turns out that `sys.modules` is only part of the problem, we would also need to register all the `numpy` objects again as `np` objects. `PythonDomain.findobj(...)` searches a dictionary of registered objects, perhaps a more fruitful direction would be to tell `findobj` that `np` is an alias for `numpy`.

Closing this, it is the wrong approach.

**Comment by eric-wieser** [comment]
I think this approach would work if you then also told sphinx to generate docs for things under `np.` instead of `numpy` - so pages like https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.dot.html would say `np.dot` at the top instead.

This might actually be a good idea from the perspective of matching source code, but worse from a search engine perspective.

**Pull request closed without merging**